### ISSUE TITLE
chain (list[int], list[str]) Unnecessary isinstance 

### ISSUE THREAD

**Issue by FrancescElies** [microsoft/pyright#4509]
**Describe the bug**
Given this reprex
```python
# mychain.py
from itertools import chain

# reportUnnecessaryIsInstance
a = ["1", "2", "3"]
b = [4, 5, 6]

for x in chain(a, b):
    if isinstance(x, int):
        print(f"int: {x}")
    elif isinstance(x, str):
        print(f"str: {x}")
    else:
        raise RuntimeError()
```
```toml
[tool.pyright]
# https://github.com/microsoft/pyright/blob/main/docs/configuration.md
typeCheckingMode = "strict"
```

**To Reproduce**
```
❯ pyright mychain.py
No configuration file found.
pyproject.toml file found at C:\s\mytest\chain.
Loading pyproject.toml file at C:\s\mytest\chain\pyproject.toml
Assuming Python version 3.11
Assuming Python platform Windows
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath C:\s\mytest\chain\typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.290
C:\s\mytest\chain\mychain.py
  C:\s\mytest\chain\mychain.py:10:10 - error: Unnecessary isinstance call; "str" is always an instance of "str" (reportUnnecessaryIsInstance)
1 error, 0 warnings, 0 informations
Completed in 1.363sec
```

**Expected behavior**
I expect no errors, or am I missing something?

**Comment by FrancescElies** [comment]
The problem is not related to chain, this function also throws an error
```python
# my_isinstance_v2.py
def my_isinstance(left: int | float) -> str:
    if isinstance(left, int):
        return "int"
    elif isinstance(left, float):
        return "float"
    else:
        raise RuntimeError
```

```
❯ pyright my_isinstance_v2.py
No configuration file found.
pyproject.toml file found at C:\s\mytest\isinstance.
Loading pyproject.toml file at C:\s\mytest\isinstance\pyproject.toml
Assuming Python version 3.11
Assuming Python platform Windows
Auto-excluding **/node_modules
Auto-excluding **/__pycache__
Auto-excluding **/.*
stubPath C:\s\mytest\isinstance\typings is not a valid directory.
Searching for source files
Found 1 source file
pyright 1.1.290
C:\s\mytest\isinstance\my_isinstance_v2.py
  C:\s\mytest\isinstance\my_isinstance_v2.py:4:10 - error: Unnecessary isinstance call; "float" is always an instance of "float" (reportUnnecessaryIsInstance)
1 error, 0 warnings, 0 informations
Completed in 0.595sec
```

**Comment by erictraut** [comment]
You appear to have enabled the strict-mode check `reportUnnecessaryIsInstance`. It's telling you that the second `isinstance` check is unnecessary because the variable must be an instance of `float` if it's not an instance of `int`. The use of unnecessary `isinstance` checks can be an indication of a logic error in your code, which is why this check was added. You can disable it if you prefer that pyright not check for it. To do so, disable "strict" mode and individually select the diagnostic checks that you want. You can also disable this particular diagnostic check on a per-file basis using a `# pyright: reportUnnecessaryIsInstance=false` at the top of the file or a `# pyright: ignore[reportUnnecessaryIsInstance]` on that line.

In any case, this isn't a bug. It is intended behavior.

**Comment by FrancescElies** [comment]
@erictraut that makes sense, I should have understood that on my own, somehow I didn't see it before, thanks for the clarification.
